### PR TITLE
Added Ngrok support

### PR DESCRIPTION
**Pull request by Sharathmk99**
## Description

Support ngrok tunneling service

## Motivation and Context

#894

## How Has This Been Tested?

```
python ./guides/quick-start/main.py
bentoml serve --run-with-ngrok IrisClassifier\20200720224248_727797
```

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] New feature and improvements (non-breaking change which adds/improves functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Code Refactoring (internal change which is not user facing)
- [ ] Documentation
- [ ] Test, CI, or build

## Component(s) if applicable
- [x] BentoService (service definition, dependency management, API input/output adapters)
- [ ] Model Artifact (model serialization, multi-framework support)
- [ ] Model Server (mico-batching, dockerisation, logging, OpenAPI, instruments)
- [ ] YataiService gRPC server (model registry, cloud deployment automation)
- [ ] YataiService web server (nodejs HTTP server and web UI)
- [ ] Internal (BentoML's own configuration, logging, utility, exception handling)
- [x] BentoML CLI

## Checklist:

- [x] My code follows the bentoml code style, both `./dev/format.sh` and
  `./dev/lint.sh` script have passed
  ([instructions](https://github.com/bentoml/BentoML/blob/master/DEVELOPMENT.md#style-check-and-auto-formatting-your-code)).
- [x] My change reduces project test coverage and requires unit tests to be added
- [ ] I have added unit tests covering my code change
- [x] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

### CODE REVIEW

**Comment by Korusuke** [comment]
@Sharathmk99 
please refer to [dev guide](https://github.com/bentoml/BentoML/blob/master/DEVELOPMENT.md#style-check-and-auto-formatting-your-code)

Run lint and format so to pass the lint tests :)

File: bentoml/utils/flask_ngrok.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    with open(download_path, 'wb') as f:
+        shutil.copyfileobj(r.raw, f)
+    return download_path
+
+
+def start_ngrok(port):
+    ngrok_address = _run_ngrok(port)
+    print(f" * Running on {ngrok_address}")
+    print(f" * Traffic stats available on http://127.0.0.1:4040")
+
```
Comment (Korusuke) [inline bentoml/utils/flask_ngrok.py:77]: In my opinion, I think we should use logging and not print statements.

File: bentoml/utils/flask_ngrok.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+def _get_command():
+    system = platform.system()
+    if system == "Darwin":
+        command = "ngrok"
+    elif system == "Windows":
+        command = "ngrok.exe"
+    elif system == "Linux":
+        command = "ngrok"
+    else:
+        raise Exception("{system} is not supported".format(system=system))
```
Comment (Korusuke) [inline bentoml/utils/flask_ngrok.py:25]: Inherit Exception from `BentoMLException`

File: bentoml/cli/bento_service.py [hunk 3]
```diff
@@ -221,9 +221,16 @@ def open_api_spec(bento=None):
         help="Run API server with micro-batch enabled",
         envvar='BENTOML_ENABLE_MICROBATCH',
     )
-    def serve(port, bento=None, enable_microbatch=False):
+    @click.option(
+        '--run-with-ngrok',
+        is_flag=True,
+        default=False,
+        help="(Beta) Run API server with ngrok enabled",
```
Comment (parano) [inline bentoml/cli/bento_service.py:228]: Nit: `help="Use ngrok to relay traffic on a public endpoint to this API server on localhost"`

File: bentoml/utils/flask_ngrok.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
+import atexit
```
Comment (parano) [inline bentoml/utils/flask_ngrok.py:1]: Any reason not to just use the flask-ngrok package directly?

File: bentoml/utils/flask_ngrok.py (hunk 1 above)
Comment (Sharathmk99) [reply bentoml/utils/flask_ngrok.py]: Fixed

File: bentoml/utils/flask_ngrok.py (hunk 2 above)
Comment (Sharathmk99) [reply bentoml/utils/flask_ngrok.py]: Fixed

File: bentoml/utils/flask_ngrok.py (hunk 4 above)
Comment (Sharathmk99) [reply bentoml/utils/flask_ngrok.py:1]: flask-ngrok package in pypi is not updated and not maintained. I tried to use flask-ngrok package directly, but latest code is not available in pypi

File: bentoml/server/__init__.py [hunk 5]
```diff
@@ -66,7 +66,7 @@ def start_dev_server(saved_bundle_path: str, port: int, enable_microbatch: bool)
         marshal_server.async_start(port=port)
         api_server.start()
     else:
-        api_server = BentoAPIServer(bento_service, port=port)
+        api_server = BentoAPIServer(bento_service, port=port, run_with_ngrok=run_with_ngrok)
```
Comment (parano) [inline bentoml/server/__init__.py]: We would also want to apply this to `enable_microbatch=True` case above.

I'm more inclined to keep ngrok related logic here within the `start_dev_server` method, and move it to the beginning of this method:

```
        if run_with_ngrok:
            from flask_ngrok import start_ngrok
            thread = Timer(1, start_ngrok, args=(port,))
            thread.setDaemon(True)
            thread.start()

          if enable_microbatch:
             ...
             api_server.start()
          else:
             ...
            api_server.start()


```

**Comment by parano** [comment]
Thanks for the PR @Sharathmk99! Just finished a pass with a few comments. There are a few linter and code formatting related errors that caused the Travis CI tests to fail, you can run `make lint` and `make format` locally and modify code style accordingly to resolve those issues.

File: bentoml/utils/flask_ngrok.py (hunk 4 above)
Comment (parano) [reply bentoml/utils/flask_ngrok.py:1]: got you, that make sense

File: bentoml/server/__init__.py (hunk 5 above)
Comment (Sharathmk99) [reply bentoml/server/__init__.py]: I agree with you. I'll make the required changes

**Comment by Sharathmk99** [comment]
> Thanks for the PR @Sharathmk99! Just finished a pass with a few comments. There are a few linter and code formatting related errors that caused the Travis CI tests to fail, you can run `make lint` and `make format` locally and modify code style accordingly to resolve those issues.

I have successfully ran both `lint`and `format`. Travis build is failing because of test case.

```
=========================== short test summary info ===========================
FAILED tests/utils/test_usage_stats.py::test_track_cli_with_keyboard_interrupt
============ 1 failed, 259 passed, 6 skipped in 363.34s (0:06:03) =============
The command "./travis/unit_tests.sh" exited with 1.
```

**Comment by parano** [comment]
@Sharathmk99 try adding the new parameter the the start_dev_server mock in that test? https://github.com/bentoml/BentoML/blob/master/tests/utils/test_usage_stats.py#L48

**Comment by Sharathmk99** [comment]
> @Sharathmk99 try adding the new parameter to the start_dev_server mock in that test? https://github.com/bentoml/BentoML/blob/master/tests/utils/test_usage_stats.py#L48

Yes it worked. How can i combine all my commits to one commit? So that it looks neat.

**Comment by parano** [comment]
@Sharathmk99 thanks for updating the PR, it looks great! You don't need to combine them, we use the "squash and merge" feature on GitHub which does that automatically. Merging now!

**Pull request merged**